src/Parallafka/CommitState.cs

```diff
@@ -0,0 +1,185 @@
+
+            this._messagesNotYetCommittedByPartitionReaderWriterLock.EnterReadLock();
+            try
+            {
+                allQueues = this._messagesNotYetCommittedByPartition.Values.ToList();
+            }
+            finally
+            {
+                this._messagesNotYetCommittedByPartitionReaderWriterLock.ExitReadLock();
+            }
+
+            foreach (var queue in allQueues)
+            {
+                IKafkaMessage<TKey, TValue> messageToCommit = null;
+
+                lock (queue)
+                {
+                    while (queue.TryPeek(out var msg) && msg.WasHandled)
+                    {
+                        if (messageToCommit == null || msg.Offset.Offset > messageToCommit.Offset.Offset)
+                        {
+                            messageToCommit = msg;
+                        }
+
+                        queue.Dequeue();
+                    }
+                }
+
+                if (messageToCommit != null)
+                {
+                    yield return messageToCommit;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Given a message ready to be committed, find a message to commit up to and including the message
+        /// </summary>
+        /// <param name="message">The message that could be committed</param>
+        /// <param name="messageToCommit">The message that should be committed</param>
+        /// <returns>True if a message was found, false otherwise</returns>
+        public bool TryGetMessageToCommit(IKafkaMessage<TKey, TValue> message, [NotNullWhen(true)] out IKafkaMessage<TKey, TValue> messageToCommit)
+        {
+            messageToCommit = null;
+            Queue<IKafkaMessage<TKey, TValue>> messagesNotYetCommitted;
+
+            this._messagesNotYetCommittedByPartitionReaderWriterLock.EnterReadLock();
+            try
+            {
+                if (!this._messagesNotYetCommittedByPartition.TryGetValue(message.Offset.Partition, out messagesNotYetCommitted))
+                {
+                    Parallafka<TKey, TValue>.WriteLine($"CS:GetMsgToCommit: {message.Key} {message.Offset} [none]");
+                    return false;
+                }
+            }
+            finally
+            {
+                this._messagesNotYetCommittedByPartitionReaderWriterLock.ExitReadLock();
+            }
+
+            lock (messagesNotYetCommitted)
+            {
+                while (messagesNotYetCommitted.TryPeek(out IKafkaMessage<TKey, TValue> msg) && 
+                       msg.WasHandled && 
+                       msg.Offset.Offset <= message.Offset.Offset)
+                {
+                    if (messageToCommit == null || msg.Offset.Offset > messageToCommit.Offset.Offset)
+                    {
+                        messageToCommit = msg;
+                    }
+
+                    messagesNotYetCommitted.Dequeue();
+
+                    this._canQueueMessage.Release();
```

Comment:
This Release() is missing in GetMessagesToCommit.

Why do both these methods have to exist in the first place?
If it's extremely necessary for both to exist, can we refactor so no logic is duplicated?

Reply:
What about modifying the commitTimer to wait for a commitAnythingReadyNowTask or something:

```
await Task.WhenAll(Task.Delay(delay, this._stopTimer.Token), tryCommitRequestedTask)
await GetAndCommitAnyMessages();
```
and have the TryCommitMessage just trigger that?

Reply:
Sometimes it takes a few looks to find the similar patterns and merge the common code....  Nice catch. Will merge logic.